Sources/Basic/Lock.swift

```diff
@@ -31,49 +31,85 @@ enum ProcessLockError: Swift.Error {
     case unableToAquireLock(errno: Int32)
 }
 
-/// Provides functionality to aquire a lock on a file via POSIX's flock() method.
+/// Provides functionality to aquire an exclusive lock on a file via POSIX's flock() method.
 /// It can be used for things like serializing concurrent mutations on a shared resource
-/// by mutiple instances of a process. The `FileLock` is not thread-safe.
+/// by mutiple instances of a process.
 public final class FileLock {
+    private typealias FileDescriptor = CInt
+
     /// File descriptor to the lock file.
-    private var fd: CInt?
+    private var fd: FileDescriptor?
+
+    private let _lock = NSLock()
```

Comment:
I don't think we need to make FileLock thread-safe. In the future, Workspace itself should implement thread safety for its operations.

Reply:
anything wrong with thread safety here?

Reply:
IMO it's just a little premature to add it unless there is a clear use case. It can manifest in weird bugs or can end up hiding real bugs since it'll kind of make Workspace thread-safe as a side-effect.

Reply:
If making a class (that is used once, in tests, and works just fine) thread-safe is a blocker, I don't know, I can only give up. Unless there is an actual problem spotted.

Reply:
It's not about how much it's being used right now. It's about what kind of abstraction we want to design for flock. In many cases, thread-safety will need to be implemented at a higher-level than FileLock anyway and then we end up with an unnecessary lock. Since this class abstracting `flock`, maybe we should see if flock itself is thread-safe. From my very little search, it doesn't look like it is explicitly marked thread-safe.

Reply:
normally this kind of access if thread safe, if you look at foundation too. since this is protecting access to the resource, I find lock appropriate to use.

Reply:
It doesn't seem thread-safe in Foundation either: https://developer.apple.com/documentation/foundation/nsdistributedlock?language=objc

Reply:
it is internally though, if you check sources

Reply:
I removed NSLock hoping this was the only obstacle.